lightning/src/util/events.rs

```diff
@@ -27,6 +27,45 @@ use prelude::*;
 use core::time::Duration;
 use core::ops::Deref;
 
+/// Some information provided on receipt of payment depends on whether the payment received is a
+/// spontaneous payment or a "conventional" lightning payment that's paying an invoice.
+#[derive(Clone, Debug)]
+pub enum PaymentPurpose {
+	/// Information for receiving a payment that we generated an invoice for.
+	InvoicePayment {
+		/// The preimage to the payment_hash, if the payment hash (and secret) were fetched via
+		/// [`ChannelManager::create_inbound_payment`]. If provided, this can be handed directly to
+		/// [`ChannelManager::claim_funds`].
+		///
+		/// [`ChannelManager::create_inbound_payment`]: crate::ln::channelmanager::ChannelManager::create_inbound_payment
+		/// [`ChannelManager::claim_funds`]: crate::ln::channelmanager::ChannelManager::claim_funds
+		payment_preimage: Option<PaymentPreimage>,
+		/// The "payment secret". This authenticates the sender to the recipient, preventing a
+		/// number of deanonymization attacks during the routing process.
+		/// It is provided here for your reference, however its accuracy is enforced directly by
+		/// [`ChannelManager`] using the values you previously provided to
+		/// [`ChannelManager::create_inbound_payment`] or
+		/// [`ChannelManager::create_inbound_payment_for_hash`].
+		///
+		/// [`ChannelManager`]: crate::ln::channelmanager::ChannelManager
+		/// [`ChannelManager::create_inbound_payment`]: crate::ln::channelmanager::ChannelManager::create_inbound_payment
+		/// [`ChannelManager::create_inbound_payment_for_hash`]: crate::ln::channelmanager::ChannelManager::create_inbound_payment_for_hash
+		payment_secret: PaymentSecret,
+		/// This is the `user_payment_id` which was provided to
+		/// [`ChannelManager::create_inbound_payment_for_hash`] or
+		/// [`ChannelManager::create_inbound_payment`]. It has no meaning inside of LDK and is
+		/// simply copied here. It may be used to correlate PaymentReceived events with invoice
+		/// metadata stored elsewhere.
+		///
+		/// [`ChannelManager::create_inbound_payment`]: crate::ln::channelmanager::ChannelManager::create_inbound_payment
+		/// [`ChannelManager::create_inbound_payment_for_hash`]: crate::ln::channelmanager::ChannelManager::create_inbound_payment_for_hash
+		user_payment_id: u64,
+	},
+	/// Because this is a spontaneous payment, the payer generated their own preimage rather than us
+	/// (the payee) providing a preimage.
+	SpontaneousPayment(PaymentPreimage),
```

Comment:
@jkczyz I was thinking that people might want to specifically send either a keysend spontaneous payment or a non-keysend spontaneous payment if they know that the payee only supports one or the other. So, I'm somewhat in favor of switching all `Spontaneous` variable names to `Keysend`. Thoughts?

Reply:
My understanding was that Keysend is a specific type of spontaneous payment that is the predecessor to spontaneous AMP payments. They both seem either unintuitive or somewhat overloaded names and also are implementation details from a user's perspective.

On the receiving side (`Event:: PaymentReceived`), does the user need to know what implementation was used to make the spontaneous payment? If not, I'd argue we shouldn't expose the implementation detail in our public API.

On the sending side (`ChannelManager`), how does the user know if the receiver accepts spontaneous payments and using which implementation? If via feature flags, could we chose which implementation to use based on the flags rather than making the user choose?

On the processing side, can these be generalized in any way?

I don't have definitive answers to these, but I'd imagine we could at least keep the first two generic (i.e., "spontaneous") and if needed make the third specific to the implementation.

Reply:
w/r/t:
> On the receiving side (Event:: PaymentReceived), does the user need to know what implementation was used to make the spontaneous payment?
> If via feature flags, could we chose which implementation to use based on the flags rather than making the user choose?

Not sure, pinged Sphinx for their thoughts last night. Hopefully they get back to me soonish

> how does the user know if the receiver accepts spontaneous payments and using which implementation?

It's a little annoying right now... C-Lightning and Eclair advertise feature bit 55 if they support keysend, but iiuc lnd doesn't advertise any feature bits for keysend. So, right now you know if the receiver *doesn't* support keysend if they fail back the payment, but there's no good way to tell if they *do* support keysend.